packages/muze-utils/src/DataSystem/selection.js

```diff
@@ -42,61 +26,54 @@ class Selection {
      * @return {Selection} Modified selection.
      * @memberof Selection
      */
-    data (newData, idGetter) {
-        if (idGetter) {
-            this._data = [];
-            this._idGetter = idGetter;
-            const tempMap = {};
+    data (newData) {
+        const entryData = new Map();
+        const exitdata = new Map();
+        const tempMap = new Map();
+        const duplicateData = new Map();
 
+        if (!this._idGetter) {
+           // when id-getter is not given
+        } else {
             newData.forEach((...params) => {
-                const index = idGetter(...params);
-                tempMap[index] = params[0];
-            });
-            // check if any data items have been removed
-            const purgedIds = [];
-
-            Object.keys(this._idMap).forEach((id) => {
-                if (!tempMap[id]) {
-                    purgedIds.push(id);
-                }
-            });
-
-            Object.keys(tempMap).forEach((id) => {
-                if (!this._idMap[id]) {
-                    this._enterdata.push(tempMap[id]);
+                const key = this._idGetter(...params);
+                if (!tempMap.has(key)) {
+                    tempMap.set(key, params[0]);
                 } else {
-                    this._idMap[id] = tempMap[id];
-                    this._data.push(tempMap[id]);
+                    duplicateData.push(key, params[0]);
                 }
             });
-            // move the purged items to exit selection
-            purgedIds.forEach((id) => {
-                const purged = this._idMap[id];
-                this._exitdata.push(purged);
-                // this._data = this._data.slice(id, 1);
-                delete this._idMap[id];
-            });
-            // this._data = this._data.slice(temp, this._data.length);
-
-            return this;
         }
-        // no id getter supplied so use indices
-        if (newData.length > this._data.length) {
-            const startIdx = this._data.length;
-            for (let i = startIdx; i < newData.length; i += 1) {
-                this._enterdata.push(newData[i]);
+
+        let val;
+        // prepare enter data
+        let entries = tempMap.entries();
+        while (val = entries.next().value) {
+            if (!this._updatedata.has(val[0])) {
```

Comment:
Store `this._updatedata` in variable

Reply:
why ??

Reply:
so that it is not accessed multiple times in a loop

Reply:
Has this been addressed?